lib/battle/BattleHexArray.h

```diff
@@ -0,0 +1,321 @@
+	using StorageType = std::vector<BattleHex>;
+
+	using value_type = BattleHex;
+	using size_type = StorageType::size_type;
+	using reference = value_type &;
+	using const_reference = const value_type &;
+	using pointer = value_type *;
+	using const_pointer = const value_type *;
+	using difference_type = typename StorageType::difference_type;
+	using iterator = typename StorageType::iterator;
+	using const_iterator = typename StorageType::const_iterator;
+	using reverse_iterator = typename StorageType::reverse_iterator;
+	using const_reverse_iterator = typename StorageType::const_reverse_iterator;
+
+	using ArrayOfBattleHexArrays = std::array<BattleHexArray, GameConstants::BFIELD_SIZE>;
+
+	static const ArrayOfBattleHexArrays neighbouringTilesCache;
+	static const std::map<BattleSide, ArrayOfBattleHexArrays> neighbouringTilesDblWide;
+
+	BattleHexArray() noexcept
+	{
+		internalStorage.reserve(totalSize);
+	}
+
+	template <typename Container, typename = std::enable_if_t<
+		std::is_convertible_v<typename Container::value_type, BattleHex>>>
+		BattleHexArray(const Container & container) noexcept
+			: BattleHexArray()
+	{
+		for(auto value : container)
+		{
+			insert(value); 
+		}
+	}
+
+	void resize(size_type size)
+	{
+		clear();
+		internalStorage.resize(size);
+	}
+	
+	BattleHexArray(std::initializer_list<BattleHex> initList) noexcept;
+
+	/// returns all tiles, unavailable tiles will be set as invalid
+	/// order of returned tiles matches EDir enum
+	static BattleHexArray getAllNeighbouringTiles(BattleHex hex)
+	{
+		static ArrayOfBattleHexArrays cache;
+		static bool initialized = false;
+
+		if(initialized)
+			return cache[hex.hex];
+
+		for(BattleHex h = 0; h < GameConstants::BFIELD_SIZE; h.hex++)
+		{
+			cache[h].resize(6);
+
+			for(auto dir : BattleHex::hexagonalDirections())
+				cache[h].set(dir, h.cloneInDirection(dir, false));
+		}
+		initialized = true;
+
+		return cache[hex.hex];
+	}
+
+	void checkAndPush(BattleHex tile)
+	{
+		if(tile.isAvailable() && !contains(tile))
+		{
+			presenceFlags[tile] = 1;
+			internalStorage.emplace_back(tile);
+		}
+	}
+
+	void insert(BattleHex hex) noexcept
+	{
+		/*if(isNotValidForInsertion(hex))
+			return;*/
+
+		if(contains(hex))
+			return;
+
+		presenceFlags[hex] = 1;
+		internalStorage.emplace_back(hex);
+	}
+
+	void set(size_type index, BattleHex hex)
+	{
+		/*if(isNotValidForInsertion(hex))
+			return;*/
+
+		if(index >= internalStorage.size())
+		{
+			logGlobal->error("Invalid BattleHexArray::set index parameter. It is " + std::to_string(index)
+				+ " and current size is " + std::to_string(internalStorage.size()));
+			throw std::out_of_range("Invalid BattleHexArray::set index parameter. It is " + std::to_string(index)
+				+ " and current size is " + std::to_string(internalStorage.size()));
+		}
+
+		if(contains(hex))
+			return;
+
+		presenceFlags[hex] = 1;
+		internalStorage[index] = hex;
+	}
+
+	iterator BattleHexArray::insert(iterator pos, BattleHex hex) noexcept
```

Comment:
There is no need to specify class name for methods that are defined in place
```suggestion
	iterator insert(iterator pos, BattleHex hex) noexcept
```